### ISSUE TITLE
Is it ok to reuse a single connection for several queries?

### ISSUE THREAD

**Issue by danilohgds** [oracle/node-oracledb#871]
Hi guys, 

This is more of a question than an issue itself, I've written a small code where I am running 3 separate queries using a single connection; this is done in a 120 seconds interval. 

What I have is a doubt between the following designs: 

Choice 1 : 

1. Provide pool object to separate class.
2. Use the pool objec,t get a  single connection, do all the queries with single connection object, and never close such connection for the entire life of the app.

Choice 2:
1. Provide pool object to separate class
2. Separate class will do getConnection from the pool for each of the queries, and them close such objects once the query is finished.

Can you please help me find out the best approach in this case?

**Comment by dmcghan** [comment]
@danilohgds 

> Is it ok to reuse a single connection for several queries?

Yes, absolutely. The general idea is to get a connection from a pool, use it to execute as many statements as needed to complete a transaction, then release the connection back to the pool.

> this is done in a 120 seconds interval

Can you clarify what you mean here? Do the 3 queries take 120 seconds total or are they executed every 120 seconds?

Can you tell us more about the app? How are the three queries related and when/why are they run? Is this an API and the three queries are needed to get some data for the response?

> Use the pool objec,t get a single connection, do all the queries with single connection object, and never close such connection for the entire life of the app.

This sounded good until the last part "never close such connection for the entire life of the app". Why not release the connection after executing the queries? You could get a new connection when you need to execute them again, no?

> Separate class will do getConnection from the pool for each of the queries

On the surface, this sounds nice because you'll be running the queries in parallel. Whether or not it's a good idea depends on your app, database, etc. For example, if your app needs transaction control, you will not really have it because you'll have three separate connections/sessions. 

Also, at some point, asking the database to do too much work at the same time starts to work against you. It seems counter-intuitive, but sometimes doing things sequentially can actually be more performant and scale better. Where this starts to happen depends on a number of factors including your hardware (CPUs, memory, discs), the load on the system, etc.

**Comment by danilohgds** [comment]
@dmcghan ,

Dan the interval means that every 120s, the application will run the same number of queries,(3).

The two queries that I run initially bring data about products that I must validate. They run every 120s in case the user has inserted a new product in the db.

I may consider using result_cache hint to increase performance, since they are invoked quite a lot.
Final query is a procedure call.

I ended up writing the code this way : 

 ```
oraclePool.getConnection().then((connection: any) => {
            connection.execute("select syntax 1", {}, {
                outFormat: oracledb.OBJECT
            }).then((data: any) => {
                if (data.rows[0] !== null && data.rows[0].FLAG == '1') {                        
                    intervalPauser$.next(true);
                } else {
                    intervalPauser$.next(false);
                }
            })
        });
```
then the other two connections, I did another pool.get connection and reused it on two queries.

 ```
let assets = connection.execute("select syntax2", {}, {
                            outFormat: oracledb.OBJECT
                        })
                        .catch((err:any) => {
                                connection.close ();
                        });
        
                        let assetsFamilies = connection.execute("select syntax3", {}, {
                            outFormat: oracledb.OBJECT
                        })
                        .catch((err:any) => {
                                connection.close ();
                        });
```

At the end I ended up wrapping these up with a promise.all into another  final query, reusing the same connection.


```
Promise.all([assets, assetsFamilies]).then(promiseReturn => { 
 connection.execute(plSqlCommand, procParams)
                                .then((result:any) => {
                                    connection.close();                           
                                    return result;
                                })
                                .catch((error:any) => {
                                    connection.close();                        
                                });
}
```

All the queries used above are not heavy, so I was only worried if using the same connection twice would be ok, or if that would cause any locks to happen. Also I added connection closures where it made sense.

My uncertainty comes down to an issue we currently have in the app where when queries in our backend were taking long, everything else would start to fall apart, even HTTP requests would start timing out, and we have no other options but to restart the container. 

As long as reusing the connection to a single run in the 120s scope is fine, then I'm ok with the code I have.

**Comment by dmcghan** [comment]
@danilohgds 

Okay, so here's the flow at a high level...

- get conn1 to determine whether to proceed with more work, close conn1
- if proceed, get conn2 to execute 2 queries with Promise.all, then use conn2 to execute PL/SQL proc, close conn2

Do I have that right?

I do see some problems here...

1. A connection can only do one thing at a time. So although you're trying to use Promise.all to do things in parallel, the queries will be executed sequentially. 
2. With Promise.all, you lose control of your code. If `select syntax2` generates an error, what will run next? The `connection.close` or the `select syntax3`? The answer would depend on various queues that you don't control.

My recommendation would be to write, at least the conn2 portion of the code, as a single promise chain (or better yet, rewrite it using an async function). Just do `select syntax2`, then `select syntax3`, then execute the PL/SQL. You would probably only need a single `close` call with this.

We'll be talking about async patterns in tomorrow's office hours session if you want to join:
https://devgym.oracle.com/pls/apex/dg/office_hours/727

> data.rows[0] !== null

I think this should be `data.rows[0] !== null` or just `data.rows.length`.

**Comment by danilohgds** [comment]
Hi Dan @dmcghan, that does make sense! I would've done this with async, however the main problem is that I am restricted to node 6 capabalities... so no async yet. 

I will join thanks for sharing the meeting.  

You can go ahead and close this for now, I think I have enough to work on already, thanks :)

**Comment by cjbj** [comment]
One final scenario: do the queries in a single PL/SQL block and return REF CURSORs?  This would be sequential execution.

**Comment by danilohgds** [comment]
@cjbj I don't see a need for the PL SQL block to return anything, as long as I return an OUT parameter with OK, that is enough. I am not opening any refcursors within the proc that I would need returned.

Pool statistics:
...total up time (milliseconds): 350050
...total connection requests: 46
...total requests enqueued: 15
...total requests dequeued: 0
...total requests failed: 0
...total request timeouts: 8
...max queue length: 10
...sum of time in queue (milliseconds): 480027
...min time in queue (milliseconds): 0
...max time in queue (milliseconds): 0
...avg time in queue (milliseconds): 32002
...pool connections in use: 14
...pool connections open: 14
Related pool attributes:
...poolAlias: PVDPOOL
...queueRequests: true
...queueTimeout (milliseconds): 60000
...poolMin: 0
...poolMax: 24
...poolIncrement: 1
...poolTimeout (seconds): 60
...poolPingInterval: 20
...stmtCacheSize: 30
Related environment variables:
...process.env.UV_THREADPOOL_SIZE: 12
2018-03-05T21:55:37.376Z HUBREQ HUB REQUEST FIRED
2018-03-05T21:55:37.623Z JOBS initialize Operations Process 0
unhandledRejection NJS-040: connection request timeout
unhandledRejection NJS-040: connection request timeout
unhandledRejection NJS-040: connection request timeout

One of the doubts I have right now is that the queries are timing out after some time that I let my node code running, so I may have another problem.

I ended up writing the process as follows (  I do have a restriction that I cannot go beyond node 6 )

Promise 1 = getConnection 1 + query + close connection
Promise 2 = getConnection 2 + query + close connection

Promise all ( promise1, promise 2) then call the procedure

**Comment by dmcghan** [comment]
@danilohgds I'm guessing the problem is that the first part of the logic gets a connection but doesn't release it after invoking `intervalPauser$.next`.

**Comment by danilohgds** [comment]
oraclePool.getConnection().then((connection: any) => {
                connection.execute("select flag from pvd.lock_lote where rownum =1", {}, {
                    outFormat: oracledb.OBJECT
                }).then((data: any) => {
                    connection.close()
                    .then(() => debugError('connection closed'));
                    if (data.rows[0].FLAG == '1') {                        
                        intervalPauser$.next(true);
                    } else {
                        debugJobs(`SBK-ROBOT-ONSHORE: FREE %O`, 'SUCESS');
                        intervalPauser$.next(false);
                    }
                })
                .catch((err:any) =>{
                    console.log(`error in select from lock lote query ${err}`);
                    connection.close(() => debugError('connection closed'));
                });
            })

@dmcghan  It has one and still does give me an error.

**Comment by dmcghan** [comment]
@danilohgds Was that call to `connection.close` there before or did you just add it and retest? 

You have a code path somewhere that is not releasing connections. You can tell with these two lines:
```
...pool connections in use: 14
...pool connections open: 14
```

"open" is the number connections in the pool. "in use" in the number of connections that are checked out of the pool via getConnection.

So when you go to get the next connection it can't get one because they are all in use and eventually the request times out.

**Comment by danilohgds** [comment]
@dmcghan It was there prior to me sending it, the situation stands the same,

I've been monitoring the sessions and the number increases from time to time with inactive connections starting to pile up, until I get the connection error. 

![image](https://user-images.githubusercontent.com/12349622/37005049-61072714-20b1-11e8-9dc5-1097b04dfaa0.png)

**Comment by danilohgds** [comment]
@dmcghan I've made an error when choosing whether to retrieve from cache or do a oracle connection, objects were never closed in that instance, I'm closing this, thanks for the help so far.

**Comment by jpollard-cs** [comment]
linking to https://github.com/oracle/node-oracledb/issues/867

**Comment by alichampion** [comment]
Hi, @dmcghan I'm working on a project where 30-40 requests come up within milliseconds.

And here's the flow at a high level...

- Create pools on the application bootstrap/start phase.
- For every request, we get a pool and then get the connection.
- We perform a query and close the connection.

This all works fine till a point when a lot of new connection open requests pop up connection gets timed out and the connection open takes some 15-90 seconds and sometimes a few milliseconds.

Now my belief in the above explanation is, that DB is unable to handle such an amount of connections and gets timed out but the DBA does not accept this explanation.

So what I did now, is I created a singleton class for both pools and used it, so in this case, a connection will open once and we will not close It from the code now with this implementation I know the connection will be expired or can be broken due to various reasons so singleton class should be so flexible that it can update the connection in singleton class and it's being done using `oracledb.connection.ping()`.

But in this approach, I feel that query execution is slow for each request what do you think?
And when 250+ subsequent requests lands on a single connection it gives ORA-01000 error.

Can you check how can I optimize the queries?

I have attached the `OralceDB.txt` file which holds JS code for my singleton OracleDB class.

[OracleDb.txt](https://github.com/oracle/node-oracledb/files/8505191/OracleDb.txt)

**Comment by cjbj** [comment]
@alichampion this is a very old issue.  Please open a new one.  Thanks.  Don't forget to set UV_THREADPOOL_SIZE.  Make sure you aren't opening pools unnecessarily.  Read the manual !

**Comment by alichampion** [comment]
@cjbj Thanks for your feedback, I can make you sure that we are not opening pools unnecessarily I'm using singleton class to achieve this.

**Comment by alichampion** [comment]
@cjbj I have opened new issue [here](https://github.com/oracle/node-oracledb/issues/1493#issue-1207746081)
